Screen/native account settings

yeezick opened the pull request:
This PR will:
- Create a `Settings` and `AccountSettings` screen
- `Settings` will be capable of logging the user out & navigating the user to EditProfile + AccountSettings
- `AccountSettings` comes with the creation of 2 backend API's: `confirmPassword` & `updatePassword` with the corresponding logic on the front-end
- It also includes some minor updates to functions in `helpers.js` as well as cleaning up some stale code, particularly in `App.js`

CONCERNS:
- There is a memory link in one of the useEffects within AccountSettings: `This is a no-op cannot perform on an unmounted component......`  this can be potentially handled with <a href="https://www.linkedin.com/feed/update/urn:li:activity:6933399770535260160/" target="_blank"> this LinkedIn post </a> but I haven't had the time to test it out

Thanks for reviewing this, this is a reminder that you rock!!

yeezick committed:
- create settings screen with skeletal functional such as routing to accountSettings and EditProfile
- add settings screen to sidemenu

yeezick committed:
create AccountSettings skeleton and add to navigation stack

yeezick committed:
- create UpdatePasswordForm component and render it conditionally
  - feels WET and is not functional yet

yeezick committed:
import textInput to remove navigation bug
THIS SCREEN SHOULD NOT BE ACCESSED FROM THE DRAWER, ONLY FROM SETTINGS SCREEN

yeezick committed:
- create deleteUser controller
- create confirmPassword controller
- implement DeleteModal confirming the users decision to delete their profile
- create corresponding frontend functions to call deleteUser and confirmPassword

yeezick committed:
- refine confirmPassword API
- refine updatePassword API
- add routes to backend for confirm adn update password APIs
- refine UpdatePasswordForm with intense useEffects and conditional render, this component needs a savior
- at this point the front end will correctly active the updateMyPassword button only if password has been confirmed and both passwords match

yeezick committed:
- finalize updatePassword controller to be functional
- complete the UpdatePassword sequence with new user token, updated password, and toggle off UpdatePasswordModal after a timer
- add conditional to prevent confirmPassword form being pinged on initial render
- use switch statement for rendering of ModalMessage text after submission of newPassword

yeezick committed:
- clear out dead code from App.js
- add functionality for signing a user out and corresponding redux method

yeezick committed:
rename handleLogout to handleSignOut in Settings screen

flexibleidealist reviewed (commented):
I resolved the issue with the simulator and am able to run the app! 

Weird input behavior:

- when I enter my password to change my password or to delete my profile, the first character of the password I type is automatically capitalized; I don't see why that should be the case after taking a look at the handle change; could this be some sort of default in the iPhone iOS / native for the input?

Possible issue with Sign In:
- when I created a new user and then reloaded the app and signed in with that user, I got directed to an EditProfile screen that wasn't pre-populated with my user info; the info showed up correctly in a console log. I'm not sure if that issue is specific to this branch or not; I'm attaching the error message associated with the sign in here
<img width="561" alt="Screen Shot 2022-05-25 at 10 16 07 AM" src="https://user-images.githubusercontent.com/80634627/170284895-c2af65fd-45a9-4bb6-8cc6-b8dcce5a150f.png">

Redirect after deleting profile:
- I successfully deleted my new user account, but there seemed to be no redirect after; it should probably redirect to the sign up screen

flexibleidealist commented on backend/controllers/users.js:
```diff
@@ -42,6 +42,21 @@ export const getOneUser = async (req, res) => {
   }
 }
 
```
in line 40 it looks like there's a typo: should be `console.error(error.message)`

flexibleidealist commented on native/screens/AccountSettings.jsx:
```diff
@@ -0,0 +1,272 @@
+  const [deletePassword, setDeletePassword] = useState('');
+  const [deletionStatus, setDeletionStatus] = useState('Pending');
+
+  useEffect(() => {
```
It seems like this useEffect doesn't run on successful deletion of the user profile. Maybe something else needs to be in the dependency array?

yeezick replied:
I figured out it was navigating to Landing but it wasn't closing the modal after signing the user out

flexibleidealist replied:
great! seems like an easy fix then!

yeezick committed:
Merge branch 'native-development' of https://github.com/yeezick/Bootcamper into screen/native-account-settings

yeezick committed:
fix conrmPassword controller to expect the credentials object and user id vs. just email and password

- relocate ConfirmPasswordMessage as more than 1 component uses it
- disable autoCapitalize funcitonality and add secureTextEntry prop to passwords
- add useEffect and confirmPassword status to DeleteModal

yeezick committed:
break apart AccountSettings to its own directory with files for each feature

yeezick committed:
update AccountSettings imports after breaking it down

flexibleidealist commented on native/screens/AccountSettings/UpdatePassword.jsx:
```diff
@@ -0,0 +1,161 @@
+        }
+        value={newPasswordForm.currentPassword}
+      />
+      {confirmedPassword !== null && <ConfirmedPasswordMessage status={confirmedPassword} />}
```
I got an error when I tried to update a password; it looks like `ConfirmedPasswordMessage` hasn't been imported; easy fix

yeezick committed:
- convert prettierrc to json so linting works properly in backend folder; lints controllers/users.js
- update singin/signup controllers to use 'userID' to avoid breaking verify endpoint
- update confirmPassword controller to prevent server outage due to missing email
- add conditional to SideMenu requiring a token to exist in SecureStore before calling verify endpoint, another app breaking issue
- convert DeleteModal to a DeleteForm
- homestretch for delete functionality which will wipe the user from Securestore and redux before navigating to landing
- add edge case handling for logged out users going to UserProfile, just for devs
- update front end api functions to reflect updates to backend

yeezick committed:
add timeout to DeleteForm dispalying success modal before navigating to landing

yeezick committed:
update deletionStatus to set to null instead of pending to close modal

yeezick committed:
Merge branch 'native-development' into screen/native-account-settings

flexibleidealist commented:
I have tested this branch again and I was able to create a new user, edit the profile, update the password, and delete the profile. I think everything looks good on this branch and it's ready to merge

flexibleidealist reviewed (approved):
Looks good; I have tested the branch again and was able to run through all the functionality!

yeezick committed:
Merge branch 'native-development' into screen/native-account-settings
